### DISCUSSION TITLE
[Question] How to manage states correctly and efficiently?

### DISCUSSION THREAD

**Discussion by luozhouyang** [reflex-dev/reflex#3392] [Q&A] [answered]
I hava three questions about how to use states  management in Reflex:
* If the websocket reconnects, will the state be restored automatically? 
* If I deploy multiple pods in a k8s cluster, how do I store state correctly?
* Using Redis as a state manager will consume a lot of memory when there are a large number of users. Is there any other solution that saves memory and has high performance?

**Comment by picklelo** [comment, marked as answer]
> If the websocket reconnects, will the state be restored automatically?

Yes, each user tab has a token that maintains a state on the backend. So even through reconnects / refreshes the state should maintain. See the [architecture post](https://reflex.dev/blog/2024-03-21-reflex-architecture/) for more info.

> If I deploy multiple pods in a k8s cluster, how do I store state correctly?

You will want to use Redis in this case, and set the `redis_url` in the `rxconfig.py` or as an environment variable. This will use Redis as the state manager instead of the in-memory one. [See here](https://github.com/reflex-dev/reflex/tree/main/docker-example#more-robust-deployment) - we need to link this in our main website docs.

> Using Redis as a state manager will consume a lot of memory when there are a large number of users. Is there any other solution that saves memory and has high performance?

How many users are you expecting? How large is your state? The states expire after 1 hour so it won't accumulate, generally it shouldn't take too much memory in Redis. At the end of the day, we will have to store the same data whether it's in Redis or another store.
